Install kernel packages before pinning

hanwen-cluster opened the pull request:
### Description of changes
1. Move DisableNouveau, InstallAdditionalKernelPackages from `update_and_reboot.yaml` to `parallelcluster.yaml`. This improves consistency of build between with and without `UpdateOsPackages`
2. Remove InstallEfiBootManager because the bug it was trying to work around already got resolved.
3. Move RHEL minor version pinning from `InstallPrerequisite` to the beginning of `PinVersion` for the following reasons
3.1 Even if we make mistakes pinning packages without installing, with minor version pinned, we will pin to the latest within the minor
3.2 `yum install -y yum-plugin-versionlock` will install the version within the minor
3.3 In terms of code style, RHEL minor version pinning fits in `PinVersion` better than `InstallPrerequisite`
4. For RHEL/Rocky, install kernel-headers, kernel-devel, kernel-devel-matched(RHEL/Rocky >=9) before pinning the version. For Ubuntu, install `linux-headers` before pinning the version
5. Add a Reboot after installing and pinning kernel packages

### Tests
* Build image on all OS has succeeded

### References
* This PR improves https://github.com/aws/aws-parallelcluster/pull/6771
* Corresponding cookbook PR https://github.com/aws/aws-parallelcluster-cookbook/pull/2972

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

himani2411 commented on cli/src/pcluster/resources/imagebuilder/update_and_reboot.yaml:
```diff
@@ -97,52 +97,6 @@ phases:
-                update-grub
-              fi
-
-      - name: DisableUnattendedUpgrades
```
IF we move this to parallelcluster.yaml it will disable any updates cloud-init does on a cron basis. This is a change in default behavior of which needs a team discussion. I am not sure if cloud-init does these updates daily either!

hanwen-cluster replied:
Yes, it is change of behavior in `build-image` command in custom images. It is not a change of behavior in official AMIs.
We will discuss

himani2411 commented on cli/src/pcluster/resources/imagebuilder/parallelcluster.yaml:
```diff
@@ -191,7 +191,45 @@ phases:
+
+                if [[ ${!OS} == "rocky8" ]] || [[ ${!OS} == "rocky9" ]] ; then
+                  if [[ ${!OS} == "rocky8" ]] ; then
+                    REPOSITORY="BaseOS"
```
We do something similar in cookbook https://github.com/aws/aws-parallelcluster-cookbook/blob/7b30a79ac6ecf2d7cd6fa7b9f1102c068bb4b73c/cookbooks/aws-parallelcluster-platform/resources/install_packages/install_packages_rocky8.rb#L33-L43. Can we check both the URL for Base and appstream?

hanwen-cluster replied:
This is the doc saying rhel8 uses baseos, rhel9 uses appstream
https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/considerations_in_adopting_rhel_9/assembly_changes-to-packages_considerations-in-adopting-rhel-9

himani2411 commented on cli/src/pcluster/resources/imagebuilder/parallelcluster.yaml:
```diff
@@ -202,9 +240,41 @@ phases:
+                update-grub
+              fi
+
+      - name: DisableUnattendedUpgrades
```
I am not in favor of moving this if we are not sure whether this affects the packages or not. 

Even if it does it will have no effect as we pin the kernel version now.

hanwen-cluster replied:
You are right. The logic of DisableUnattendedUpgrades had already existed in `parallelcluster.yaml`. We need the logic in both `parallelcluster.yaml` and `update_and_reboot.yaml` to make sure unattended upgrade does not happen in any stages. 

I reverted this change

hanwen-cluster committed:
Install kernel packages before pinning

1. Move DisableNouveau, InstallAdditionalKernelPackages from `update_and_reboot.yaml` to `parallelcluster.yaml`. This improves consistency of build between with and without `UpdateOsPackages`
2. Remove InstallEfiBootManager because the bug it was trying to work around already got resolved.
3. Move RHEL minor version pinning from `InstallPrerequisite` to the beginning of `PinVersion` for the following reasons
3.1 Even if we make mistakes pinning packages without installing, with minor version pinned, we will pin to the latest within the minor
3.2 `yum install -y yum-plugin-versionlock` will install the version within the minor
3.3 In terms of code style, RHEL minor version pinning fits in `PinVersion` better than `InstallPrerequisite`
4. For RHEL/Rocky, install kernel-headers, kernel-devel, kernel-devel-matched(RHEL/Rocky >=9) before pinning the version. For Ubuntu, install `linux-headers` before pinning the version
5. Add a Reboot after installing and pinning kernel packages
